File: backend/app/schemas/analysis.py
from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, Field, StringConstraints

from app.schemas.common import ORMModel
from app.schemas.product import ExtractedProductDataResponse
from app.schemas.simulation import ICPProfileResponse, ScenarioResponse, SimulationRunResponse

WorkflowStage = Literal[
    "product_understanding",
    "icp_profiles",
    "scenarios",
    "decision_flow",
    "final_review",
]
WorkflowStepStatus = Literal["not_started", "processing", "awaiting_review", "completed", "failed", "stale"]


class AnalysisCreateRequest(BaseModel):
    url: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]
    force_refresh: bool = False
    run_async: bool = True


class WorkflowStepResponse(BaseModel):
    stage: WorkflowStage
    label: str
    status: WorkflowStepStatus
    is_current: bool
    is_complete: bool
    started_at: datetime | None = None
    completed_at: datetime | None = None
    edited: bool = False
    error_message: str | None = None


class AnalysisWorkflowResponse(BaseModel):
    current_stage: WorkflowStage
    next_stage: WorkflowStage | None = None
    steps: list[WorkflowStepResponse] = Field(default_factory=list)
    available_actions: list[str] = Field(default_factory=list)


class AnalysisListItemResponse(ORMModel):
    id: str
    input_url: str
    normalized_url: str
    status: str
    current_stage: WorkflowStage
    created_at: datetime
    completed_at: datetime | None
    error_message: str | None


class AnalysisCreateResponse(BaseModel):
    analysis: AnalysisListItemResponse
    reused: bool = False
    cloned_from_analysis_id: str | None = None


class AnalysisDetailResponse(ORMModel):
    id: str
    input_url: str
    normalized_url: str
    status: str
    current_stage: WorkflowStage
    started_at: datetime | None
    completed_at: datetime | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime
    workflow: AnalysisWorkflowResponse
    extracted_product_data: ExtractedProductDataResponse | None = None
    icp_profiles: list[ICPProfileResponse] = Field(default_factory=list)
    scenarios: list[ScenarioResponse] = Field(default_factory=list)
    simulation_runs: list[SimulationRunResponse] = Field(default_factory=list)


class WorkflowProceedRequest(BaseModel):
    expected_stage: WorkflowStage
    run_async: bool = True


class WorkflowReopenRequest(BaseModel):
    stage: Literal["product_understanding", "icp_profiles", "scenarios"]
    entity_id: str | None = None

File: frontend/src/lib/analysis.ts
import { AnalysisDetail, AnalysisListItem, ICPProfile, Scenario, SimulationRun, WorkflowStage } from "@/types/api";

const STAGES: WorkflowStage[] = [
  "product_understanding",
  "icp_profiles",
  "scenarios",
  "decision_flow",
  "final_review",
];

const STAGE_LABELS: Record<WorkflowStage, string> = {
  product_understanding: "Product Understanding",
  icp_profiles: "ICP Profiles",
  scenarios: "Suggested Scenarios",
  decision_flow: "Decision Flow",
  final_review: "Final Review",
};

export function createPendingAnalysisDetail(analysis: AnalysisListItem): AnalysisDetail {
  return {
    id: analysis.id,
    input_url: analysis.input_url,
    normalized_url: analysis.normalized_url,
    status: analysis.status,
    current_stage: analysis.current_stage,
    started_at: null,
    completed_at: analysis.completed_at,
    error_message: analysis.error_message,
    created_at: analysis.created_at,
    updated_at: analysis.created_at,
    workflow: {
      current_stage: analysis.current_stage,
      next_stage: nextStage(analysis.current_stage),
      steps: STAGES.map((stage) => ({
        stage,
        label: STAGE_LABELS[stage],
        status:
          stage === analysis.current_stage
            ? analysis.status === "queued"
              ? "not_started"
              : "processing"
            : "not_started",
        is_current: stage === analysis.current_stage,
        is_complete: false,
        started_at: null,
        completed_at: null,
        edited: false,
        error_message: null,
      })),
      available_actions: [],
    },
    extracted_product_data: null,
    icp_profiles: [],
    scenarios: [],
    simulation_runs: [],
  };
}

export function upsertAnalysisListItem(
  analyses: AnalysisListItem[] | undefined,
  analysis: AnalysisListItem,
): AnalysisListItem[] {
  const current = analyses ?? [];
  const withoutMatch = current.filter((item) => item.id !== analysis.id);
  return [analysis, ...withoutMatch];
}

export function getLatestRunForScenario(analysis: AnalysisDetail, scenarioId: string): SimulationRun | undefined {
  return [...analysis.simulation_runs]
    .filter((run) => run.scenario_id === scenarioId)
    .sort((left, right) => new Date(right.created_at).getTime() - new Date(left.created_at).getTime())[0];
}

export function getScenarioById(analysis: AnalysisDetail, scenarioId: string): Scenario | undefined {
  return analysis.scenarios.find((scenario) => scenario.id === scenarioId);
}

export function getICPById(analysis: AnalysisDetail, icpId: string): ICPProfile | undefined {
  return analysis.icp_profiles.find((icp) => icp.id === icpId);
}

export function scenarioSummaries(analysis: AnalysisDetail) {
  return analysis.scenarios.map((scenario) => ({
    scenario,
    run: getLatestRunForScenario(analysis, scenario.id),
  }));
}

export function nextStage(stage: WorkflowStage): WorkflowStage | null {
  const index = STAGES.indexOf(stage);
  if (index === -1 || index >= STAGES.length - 1) return null;
  return STAGES[index + 1];
}
